File: utils/python/resource_client.py
import os
from datetime import datetime
from typing import List, Union
from utils.python.db_client import NotionClient
from utils.python.social_client import SocialClient

class ResourceClient(SocialClient):
    def __init__(self, source: str, dateFormat: str) -> None:
        super().__init__()

        self.db = NotionClient()
        self.source = source 
        self.dateFormat = dateFormat

        # flag to refetch the entire posts or just check for new posts
        # if source is not added yet or manually enter
        self.refetch = int(os.environ.get("REFETCH", 0)) == 1
        self.new_source = not self.db.sourceExists(source)
        self.refetch = self.refetch or self.new_source
        self.delete = int(os.environ.get("DELETE", 0)) == 1
    
    def formatTitle(self, title: str) -> str:
        return title.strip()

    def formatURL(self, url: str) -> str:
        return url.strip()

    def formatPublishedOn(self, date: str) -> str:
        if date is None or self.dateFormat is None:
            return date

        date = date.strip()
        publishedOn = date
        try:
            _datetime = datetime.strptime(date, self.dateFormat)
            publishedOn = _datetime.isoformat()
        except:
            print("Invalid date format")

        return publishedOn

    def formatAuthors(self, authors: Union[str, List[str], None]):
        if authors is None:
            return authors
        
        if type(authors) is str:
            return authors.strip()
        return authors
        
    def formatTags(self, tags: Union[str, List[str], None] = None) -> str:
        if tags is None:
            return tags
        
        if type(tags) is str:
            tags = tags.strip().split(',')
        return tags

File: utils/js/resource_client.ts
import { Logger, createLogger, format, transports } from "winston";
import { DBClient } from "utils/js/db_client";
import date from "date-and-time";

class ResourceClient {
  title: string;
  url: string;
  icon: string;
  dateFormat: string;
  logger: Logger;
  db: DBClient;
  source_id: string;
  refetch: boolean;

  constructor(title: string, url: string, icon: string, dateFormat: string) {
    this.logger = createLogger({
      level: "info",
      format: format.combine(
        format.timestamp({ format: "YYYY-MM-DD HH:mm:ss" }),
        format.printf(({ level, message, label, timestamp }) => {
          return `[${level}]:[${timestamp}]:[${title}] ${message}`;
        })
      ),
      transports: [new transports.File({ filename: "resource_scrapper.log", level: "info" })],
    });

    this.db = new DBClient("database.db", this.logger);

    this.title = title;
    this.url = url;
    this.icon = icon;
    this.dateFormat = dateFormat;

    this.db.handleSource(this.title, this.url, this.icon);

    this.source_id = this.db.getSourceId(title);
    this.refetch = process.env.REFETCH == "1";
  }

  formatTitle(title: string) {
    return title.trim();
  }

  formatURL(url: string) {
    return url.trim();
  }

  formatPublishedOn(date_str?: string) {
    if (!date_str) return date_str;

    date_str = date_str.trim();
    let publishedOn = date_str;
    try {
      publishedOn = date.parse(date_str, this.dateFormat).toUTCString();
    } catch (err) {
      console.log("Invalid dateformat");
    }
    return publishedOn;
  }

  formatAuthors(authors?: string | Array<string>) {
    if (!authors) return authors;
    if (typeof authors === "string") return authors.trim();
    return authors.map((author) => author.trim()).join(",");
  }

  formatTags(tags?: string | Array<string>) {
    if (!tags) return tags;
    if (typeof tags === "string") return tags.trim();
    return tags.map((tag) => tag.trim()).join(",");
  }
}

export { ResourceClient };